packages/foam-core/src/note-graph.ts

```diff
@@ -24,98 +49,112 @@ export interface NoteLinkDefinition {
   position?: Position;
 }
 
-export class Note {
-  public id: ID;
-  public frontmatter: object;
-  public title: string | null;
-  public source: string;
-  public path: string;
-  public start: Point;
-  public end: Point;
-  public eol: string;
-  public links: NoteLink[];
-  public definitions: NoteLinkDefinition[];
-
-  constructor(
-    id: ID,
-    frontmatter: object,
-    title: string | null,
-    links: NoteLink[],
-    definitions: NoteLinkDefinition[],
-    start: Point,
-    end: Point,
-    path: string,
-    source: string,
-    eol: string
-  ) {
-    this.id = id;
-    this.frontmatter = frontmatter;
-    this.title = title;
-    this.source = source;
-    this.path = path;
-    this.links = links;
-    this.definitions = definitions;
-    this.start = start;
-    this.end = end;
-    this.eol = eol;
-  }
-}
-
 export type NoteGraphEventHandler = (e: { note: Note }) => void;
 
+export type NoteQuery = {
+  id?: ID;
+  slug?: string;
+  uri?: URI;
+  title?: string;
+}
+
 export class NoteGraph {
   private graph: Graph;
   private events: EventEmitter;
+  private createIdFromURI: (uri: URI) => ID;
 
   constructor() {
     this.graph = new Graph();
     this.events = new EventEmitter();
+    this.createIdFromURI = hashURI;
```

Comment:
Any benefits in hashURI vs UUID? Just thinking that UUID would be more future proof when merging multiple foams and maybe even reusing their cache files. With hashURI uids might collide, with UUID most probably not.

Reply:
the major advantage of basing the ID of the URI, instead of using a UUID, is shown in line 90 of this file.
```
const targetId = this.createIdFromURI(targetPath);
```
If we use the UUID we wouldn't be able to generate a connection at this stage, unless the target has already been added to the graph and we would still need to use the URI to identify it.

Even in that case, given that we must assume that docs will have links to other docs that have not been added to the graph yet, we would need to add an explicit linking phase.

I don't have a huge issue around this, but it does make the loading of the graph more complicated, so I am thinking of waiting to make such a change until there is a tangible advantage in doing that

Reply:
maybe this is good enough for now; just want to point out that this makes ids now non-opaque (?), since we have semantics associated with them (the requirement we discussed earlier). In the long run I'd really like to see distinct linking phase after the full workspace load.